### ISSUE TITLE
Phase 6 research: add objective-consistent Bernoulli/Poisson numerics for future experimental lineage

### ISSUE THREAD

**Issue by H-T-0726** [H-T-0726/expfam-lsm-attributed-relational-data#25]
## Purpose

Follow up the merged Issue #23 / PR #24 audit, which found:

- **PC-001 HIGH**: Poisson hard clipping makes the reported objective locally flat outside the clip interval while score / precision remain nonzero.
- **PC-002 MEDIUM**: Bernoulli probability and curvature floors make objective / score / precision describe different tail behavior.

The goal of this Issue is to create a **forward-only, objective-consistent numerical lineage** for future fixed/masked/per-column research work, without rewriting the historical implementation files that generated earlier results.

This Issue does **not** claim that per-column families are empirically useful. It only removes the numerical objective/derivative inconsistency identified by the audit so that a later validation experiment can be scientifically interpretable.

---

## Current main / provenance note

Start from the actual latest `origin/main`.

At Issue creation time the latest main is expected to be:

`ded896752cb3333bdc1aabbb63fdd78dda50411d`

Immediately after PR #24 the merge commit was:

`d8ddc29f81b96b824fa6010fcbcdf3a24695468c`

Two subsequent tree-neutral cleanup commits were created while restoring an accidentally created empty root file. `git diff d8ddc29..ded8967` is expected to contain **zero file differences**. Do not rewrite history; simply branch from actual latest main.

---

## Why this must be forward-only

The same legacy numerical safeguards appear in the existing scalar/fixed/masked lineage:

- `expfam/src/model_expfam.py`
- `expfam/src/model_dual_expfam.py`
- `expfam/src/model_dual_expfam_fixed.py`
- `expfam/src/experimental/model_dual_expfam_masked.py`
- `expfam/src/experimental/model_dual_expfam_percolumn.py`

Earlier research results are tied to those historical code paths / Git SHAs. Do not silently rewrite them and then imply that old results came from the new numerics.

The historical files listed below are therefore protected in this Issue.

---

## Required design principle

For every accepted Bernoulli/Poisson parameter value, the following must describe **one mathematical objective**:

1. log-likelihood / Q contribution
2. score used by E-step / M-step
3. curvature / precision used by Laplace/Newton

No hard clipping/floor may be used in only one of these three layers.

### Bernoulli

Prefer the exact canonical Bernoulli objective evaluated stably, e.g. a `logaddexp`-based form, with score and curvature corresponding to that same objective.

Do not hard-floor probabilities in the objective while using an unfloored score.
Do not impose an independent positive curvature floor unless it is explicitly defined as a surrogate and approved separately.

### Poisson

The default design target is the canonical Poisson log-link objective:

```text
mu = exp(eta)
log p = x*eta - exp(eta) + constant
score = x - exp(eta)
curvature = exp(eta)
```

Do **not** keep the current `clip(eta, -20, 10)` behavior while continuing to use nonzero residual/curvature outside the clip.

Preferred safety policy:

- preserve the canonical objective on the representable domain;
- detect non-finite / overflow-risk eta explicitly;
- fail fast with a clear numerical exception/diagnostic rather than silently switching to a different objective.

If implementation evidence shows that a smooth surrogate/capped model is necessary instead, **STOP before implementing that semantic change and report it for human approval**. A surrogate is not to be silently called canonical Poisson.

Any numeric range threshold must be derived from dtype/runtime limits or otherwise explicitly justified; do not invent a scientific cutoff such as 10 merely for convenience.

---

## Forward lineage requirement

Do not modify these historical/core implementation files:

- `expfam/src/model_expfam.py`
- `expfam/src/model_dual_expfam.py`
- `expfam/src/model_dual_expfam_fixed.py`
- `reproduction/**`

Do not overwrite the audited legacy per-column behavior merely to make old tests pass.

Preferred architecture is a **new objective-consistent future lineage** that can be selected explicitly for new experiments while legacy behavior remains recoverable and identifiable.

A reasonable design may include new classes/helpers such as:

- an objective-consistent scalar/masked experimental class for fair baselines;
- an objective-consistent per-column class;
- a small shared numerical helper module;
- explicit selection from `experimental/em_runner.py`.

These names are examples, not mandatory. Derive the minimal architecture from the actual inheritance tree.

### Explicit selection

If `run_em_experimental` is extended, use an explicit opt-in mode such as:

```text
numerics_mode = "legacy" | "consistent"
```

or an equally clear mechanism.

Backward compatibility requirement:

- default behavior remains legacy unless there is strong repository evidence to change it;
- future validation experiments can explicitly request the consistent lineage;
- no silent behavior switch for historical scripts.

---

## NB isolation

`DualExpFamLSMNB` inherits from the masked lineage but has its own non-canonical NB2 score/curvature and its own clipping logic.

**NB is out of scope for semantic modification in this Issue.**

Required:

- legacy NB construction and existing NB tests must remain unchanged and passing;
- if the proposed inheritance design would alter NB behavior, redesign the class boundary or STOP and report the coupling;
- `numerics_mode="consistent"` with `family_y="nb"` should be explicitly rejected unless NB consistency is separately implemented and approved.

Do not accidentally claim that fixing Poisson fixes NB.

---

## Scope

Expected work:

1. inspect the complete inheritance / runner path;
2. implement minimal forward-only consistent numerics;
3. add deterministic boundary/objective-consistency tests;
4. add explicit runner selection if needed for future experiments;
5. add a dated implementation/validation note.

Suggested branch:

`experiment/25-objective-consistent-numerics`

Expected report:

`reports/per_column_family/objective_consistency_fix_20260821.md`

The exact code-file set must be justified before editing.

---

### A. Legacy preservation

Prove that protected legacy classes/files are unchanged.

At minimum:

- no diff in `model_expfam.py`
- no diff in `model_dual_expfam.py`
- no diff in `model_dual_expfam_fixed.py`
- no diff in `reproduction/**`

If an existing legacy per-column class is kept unchanged, include a control showing its historical clip/floor behavior still reproduces the Issue #23 counterexample.

### B. Bernoulli consistency

Use deterministic fixtures across:

- interior logits
- positive tail
- negative tail

Verify the stable objective, score, and curvature correspond to the same function.

Use finite differences where numerically meaningful.
For extreme curvature where finite-difference cancellation dominates, document the limitation rather than fabricating precision.

Required regression target includes the Issue #23 tail case conceptually equivalent to:

```text
eta = 30, x = 0
```

The consistent lineage must not reproduce the old `objective derivative = 0` versus `score ≈ -1` contradiction caused by hard probability flooring.

### C. Poisson consistency

Required regression target:

```text
eta = 11.5, x = 3
```

This was outside the legacy upper clip of 10 while still completely representable in float64.

For the consistent lineage, objective finite differences must agree with score and curvature within justified tolerances.

Also test:

- interior eta
- negative eta below the old `-20` threshold
- explicit overflow/non-finite guard behavior

A guard failure must be loud and deterministic; it must not silently clip and continue.

### D. X and Y paths

Check both sides where supported by the consistent lineage:

- scalar X Bernoulli / Poisson
- Y Bernoulli / Poisson
- per-column X Bernoulli / Poisson
- Gaussian paths remain unchanged/equivalent

### E. EM smoke

Run only small deterministic smoke tests.

Required:

- scalar consistent model can run a small EM fixture;
- per-column consistent model can run a small mixed-X fixture;
- no NaN / silent clipping;
- if overflow guard triggers, run must fail explicitly rather than return a misleading Q.

Do not run claim-bearing large experiments in this Issue.

### F. Runner selection

If runner opt-in is added:

- omitted/default option selects legacy behavior;
- explicit consistent option selects new classes;
- family_x_list consistent path selects the consistent per-column class;
- NB legacy path is unchanged;
- NB + consistent is explicitly rejected unless separately approved.

---

## Existing tests / regression

Run all directly relevant existing tests unchanged, including at least:

```text
python expfam/src/experimental/test_percolumn_model.py
python expfam/src/experimental/test_experimental_models.py
python expfam/src/experimental/test_diagnostics_validation.py
python expfam/src/experimental/test_nb_math_audit.py
```

Discover and run other directly relevant tests if present.

Do not weaken existing assertions to obtain PASS.

---

## Q / information criterion

The consistent lineage must continue to use the repository's existing strict-Q / Q-based criterion conventions unless a genuine incompatibility is found.

Do not rename the historical `BIC` columns/functions in this Issue.
Do not claim Schwarz BIC.

Verify that the objective used by strict-Q is the same objective whose derivatives the consistent model uses, up to family constants intentionally added by the strict-Q wrapper.

If that cannot be achieved without changing `eval_utils.py`, explain the exact reason and keep the change minimal and forward-compatible.

---

## Diagnostics

Existing `poisson_clip_diagnostics` describes legacy hard clipping.

Do not silently reuse a legacy “clip rate” diagnostic as though it describes the new consistent mode.

If `run_em_experimental` exposes consistent mode:

- legacy mode may keep existing clip diagnostics;
- consistent mode should either expose a clearly named overflow/range diagnostic or return an explicit not-applicable result;
- do not report `clip_rate=0` simply because clipping was removed.

---

## Report

Create:

`reports/per_column_family/objective_consistency_fix_20260821.md`

Include:

1. base/head SHA
2. protected historical lineages
3. new forward lineage / inheritance
4. exact Bernoulli numerical objective
5. exact Poisson numerical objective
6. overflow/fail-fast policy
7. code ↔ objective ↔ score ↔ curvature table
8. Issue #23 counterexample before/after
9. finite-difference results
10. runner selection behavior
11. NB isolation result
12. existing-test results
13. lightweight CI results
14. known limitations
15. explicit statement that no claim-bearing validation experiment was run
16. explicit statement that the per-column prototype is still not promoted to the thesis/main method

---

## Forbidden changes

Do not modify:

- `expfam/src/model_expfam.py`
- `expfam/src/model_dual_expfam.py`
- `expfam/src/model_dual_expfam_fixed.py`
- `reproduction/**`
- existing research CSV / runinfo / NPY / NPZ / figures / PDF
- historical experiment outputs
- `EXPERIMENT_REGISTRY.md`
- `RESEARCH_MASTER.md`
- `KNOWN_ISSUES.md`
- `.github/workflows/**`
- `.claude/**`
- requirements files
- `.gitignore`

No moves/deletes.

Do not start the per-column validation experiment in this Issue.

---

## Validation gate

Before commit:

- exact changed-file list justified
- protected historical files: zero diff
- research result/provenance files: zero diff
- new objective/score/curvature boundary tests: PASS
- Issue #23 Poisson counterexample: fixed in consistent mode
- Issue #23 Bernoulli counterexample: fixed in consistent mode
- legacy behavior still explicitly identifiable
- NB regression: PASS / unchanged
- existing tests: PASS
- `git diff --check`: PASS

Also run the Phase 5b lightweight commands:

```text
python -m compileall -q -x 'archive[\\/]' tools expfam/src reproduction/src
python tools/validate_registry_paths.py --self-test
python tools/validate_registry_paths.py
```

Required validator state:

- self-test all PASS
- FAIL 0
- SKIP 0
- TRUE_BROKEN 0
- registry exit 0

---

## Independent review

After implementation, perform a fresh adversarial review focused on:

- accidental semantic changes to legacy classes
- objective/score/curvature consistency
- overflow guard correctness
- finite-difference methodology
- runner default compatibility
- NB coupling
- strict-Q consistency
- historical provenance

Classify findings as BLOCKER/HIGH/MEDIUM/LOW.

BLOCKER or HIGH in the implementation/review must be resolved before Draft PR delivery or reported as a stop condition.

---

## Delivery

Proceed through:

latest-main preflight → branch → read-only design → implementation → deterministic tests → report → adversarial review → commit → normal push → **Draft PR** → remote lightweight CI → remote diff review → STOP.

Do not automatically:

- mark Ready for Review
- merge
- close Issue #25 manually
- start validation experiments
- update canonical research claims

Draft PR must clearly state:

- historical files unchanged
- legacy/default mode behavior
- consistent opt-in behavior
- PC-001 status
- PC-002 status
- NB status
- exact tests
- changed files
- research results changed: NO
- validation experiment run: NO
- prototype promoted: NO
- Merge: NO — HUMAN REVIEW REQUIRED

---

## Completion gate

This Issue is complete only after human review confirms that a future experiment can explicitly select a numerically objective-consistent scalar/per-column lineage without rewriting historical implementations or silently changing old experiment semantics.

Only after merge should a separate Issue design the actual per-column validation experiment.
